source/Uniform/InMemory/InMemoryCollection.cs

```diff
@@ -18,12 +16,12 @@ public class InMemoryCollection : IDocumentCollection, IInMemoryCollection
         /// <summary>
         /// Main data structure that contains all documents of type TDocument, hashed by key
         /// </summary>
-        private readonly Dictionary<String, Object> _documents = new Dictionary<String, Object>();
+        private readonly ConcurrentDictionary<String, Object> _documents = new ConcurrentDictionary<String, Object>();

```

Comment:
Multi-thead tests show 0.25% of performance lost when comparing `ConcurrentDictionary` with `Dictionary`, which is OK, IMHO. So I don't see any problem here.

```c#
const int threadsCount = 4;
const int loopLength = 3 * 1000 * 1000 ; // 3 millions operations for each thread
var options = new ExecutionDataflowBlockOptions { MaxDegreeOfParallelism = threadsCount };
Stopwatch sw = new Stopwatch();
 
 
// Dictionary
var abDifferentDictionaries = new ActionBlock<Dictionary<int, int>>(d =>
{
    for (var i = 0; i < loopLength; i++) d[i] = 100500;
}, options);
 
sw.Start();
for (var i = 0; i < threadsCount; i++)
    abDifferentDictionaries.Post(new Dictionary<int, int>());
 
abDifferentDictionaries.Complete();
abDifferentDictionaries.Completion.Wait();
Console.WriteLine("baseline = " + sw.Elapsed);
 
 
// Dictionary + lock
var abSingleDictionaryWithLock = new ActionBlock<Dictionary<int, int>>(d =>
{
    for (var i = 0; i < loopLength; i++) lock(d) d[i] = 100500;
}, options);
 
sw.Restart();
var d1 = new Dictionary<int, int>();
for (var i = 0; i < threadsCount; i++)
    abSingleDictionaryWithLock.Post(d1);
 
abSingleDictionaryWithLock.Complete();
abSingleDictionaryWithLock.Completion.Wait();
Console.WriteLine("baseline = " + sw.Elapsed);
 
 
// ConcurrentDictionary
var abSingleConcurrentDictionary = new ActionBlock<ConcurrentDictionary<int, int>>(d =>
{
    for (var i = 0; i < loopLength; i++) d[i] = 100500;
}, options);
 
sw.Restart();
var d2 = new ConcurrentDictionary<int, int>();
for (var i = 0; i < threadsCount; i++)
    abSingleConcurrentDictionary.Post(d2);
 
abSingleConcurrentDictionary.Complete();
abSingleConcurrentDictionary.Completion.Wait();
Console.WriteLine("baseline = " + sw.Elapsed);
 
 
// Execution result + single-thread results (threadsCount = 1) on same PC in parentheses:
// baseline = 00:00:00.3115546 (00:00:00.1503635) - Dictionary
// baseline = 00:00:00.5319161 (00:00:00.1589537) - Dictionary + lock
// baseline = 00:00:00.9960964 (00:00:00.9039555) - ConcurrentDictionary
```